### DISCUSSION TITLE
Bugs related to plugin libraries

### DISCUSSION THREAD

**Discussion by tomboylover93** [tsukasa/BdBrowser#20] [Problems and issues]
So, a while ago, I decided to update the (internal) BD version to 1.8.5, and this created a really strange bug where some of the plugins related to plugin libraries (mainly ZeresPluginLibrary) make it so themes don't load, the BetterDiscord logo is always blinking in the bottom right corner even after a long session uptime, and a lot of BD's own Settings Panel is locked.

I don't know what might be the cause for this, but I did notice that the last update to BDBrowser as an extension was 1.8.2, meanwhile BD itself is on 1.8.5, so I'm thinking that there might be some sort of conflict between the extension's files and BD's own. If so, is it possible to update the extension to fully support the latest version?

Here's a screenshot I took of the Settings tab, with the DevTools console opened up on the right side and 2 different messages warning me about "0PluginLibrary.plugin.js"'s inability to load.
![Screenshot 2023-03-06 19 06 09](https://user-images.githubusercontent.com/95257311/223247889-6e695261-dac1-4ce7-b18d-55c15db21f01.png)

**Comment by tsukasa** [comment]
Greetings,

the extension has not been updated since late November 2022 because it does not need any changes to run current versions of BetterDiscord. Updating just the version number to reflect the compatible BD version would produce unnecessary "noise" in the repository. The versioning schema is a leftover from the time when BdBrowser had to be updated with each new BetterDiscord version because it shipped a version of BD's renderer.js, and direct in-place upgrades were not a thing.

As mentioned in other recent posts, I cannot reproduce the issues people seem to have lately, so debugging and fixing them proves challenging. It is a vexing topic and could potentially have multiple causes, ranging from temporary issues on some GitHub API, browser extensions that block/alter certain behaviours, and outdated/broken plugins causing issues.

Currently, my best guess is that certain plugins might be causing issues. My recommendation is:
- Create a VFS backup through the developer console because you will nuke your VFS in the next step: `require("fs").exportVfsBackup();`
- After downloading the backup, format the virtual filesystem: `require("fs").formatVfs(true);`
- Hard reload the page (Ctrl+F5) after formatting to force an initialisation of BdBrowser.
- Ideally, BetterDiscord should load now and bring up its changelog dialogue.
- If that is not the case, you can always restore your previously created backup: `require("fs").importVfsBackup();`
- If BetterDiscord loads up, you can start adding plugins. Take note of when it starts breaking again and whether this resolves if you remove the offending plugin.

If a plugin is the cause of your issues, it would help immensely if you could isolate it, so I can add it to my test setup and see what is happening. I am only able to fix issues I am able to grasp and - ideally - reproduce.

**Reply by tomboylover93** [reply]
I decided to test with the 2 plugin libraries I had at hand, which were BDFDB and ZeresPluginLibrary. BDFDB loaded up with no errors, didn't cause any of the issues I described, and plugins that depend on it work just fine. But when I added a plugin that depends on ZeresPluginLibrary to work, which then prompted me to download the plugin library, I decided to reload the site and got _this_ on the console.
![Screenshot 2023-03-07 18 15 55](https://user-images.githubusercontent.com/95257311/223556375-5b3bd50f-62e0-4e03-b3b3-9ed2fcb0a18b.png)

This is with pretty much all of my browser extensions disabled, which gets rid of the suspicion of one of them being the root of the issue. All I had to do was add a plugin, download the library needed for it to work, then reload.
I also tried downloading the library itself off of BetterDiscord's site's "Plugins" section, then load it and do the same process, and the same thing happened.

I did notice that it seems that every time I get one of these problems, it's directly related to a library-dependant plugin, or its library, having a fluke; I've also had similar issues with BDFDB in the past.

This is the file for ZeresPluginLibrary, in case you want to try it yourself and see if you get the same issue.
https://betterdiscord.app/plugin/ZeresPluginLibrary

**Reply by tsukasa** [reply]
Greetings,

thank you so much for the insights, I believe this gives me the necessary details to troubleshoot the issue. It would be awesome if you could check whether the information below applies to your scenario and report back.

**Background**
`rename` and `renameSync` are not implemented in the VFS, which would explain why it crashes out. Move and rename are not implemented for two simple reasons:
- Strencher's original version never had them.
- Implementing the IndexedDB version of the VFS would be more complex for little to no gain since only a fringe use case (detailed below) uses them.

So what exactly is going on? I suspect your plugin is not called `myplugin.plugin.js` but something wild like `myplugin.plugin (1).js` when you upload it. BetterDiscord recognizes this discrepancy between the expected filename format and the file in upload. It tries to correct the filename, which fails because the VFS does not support rename/move operations.

**Note:**
The following two examples assume you have Zerebos' BetterRoleColors plugin waiting to be uploaded.

On a fresh installation, this test run works fine for me:
- Format the VFS to ensure we work on a clean slate without external disturbances.
- Reload Discord's site to initialize a fresh BetterDiscord.
- Go to Settings - BetterDiscord - Plugins and upload the `BetterRoleColors.plugin.js`.
- The prompt to install ZeresPluginLibrary should pop up.
- The plugin can be enabled and works.

To reproduce your error, something like this would work:
- Format the VFS to ensure we work on a clean slate without external disturbances.
- Reload Discord's site to initialize a fresh BetterDiscord.
- Go to the Settings - BetterDiscord - Plugins and upload the `BetterRoleColors.plugin (1).js`.
- The console will spit out an error:
  `[BetterDiscord] [PluginManager] Could not rename file: BetterRoleColors.plugin (1).js BetterRoleColors.plugin.js TypeError: y.renameSync is not a function`
- While the plugin appears in the plugin list, no prompt to install ZeresPluginLibrary pops up, activating the plugin results in errors and the VFS errors persist until you delete the plugin from the VFS and reload the page.
- If you do not delete the plugin but _do_ reload the page, the exact message from your stack trace will be printed during initialization and BetterDiscord is stuck.

**Workaround:**
Please ensure the plugin and theme names conform to the expected format before uploading them into BetterDiscord through BdBrowser.

In case you are "stuck" on with no option to delete the plugin via BetterDiscord's Plugins tab (due to BetterDiscord not initializing properly anymore), you can list the plugins/themes via the developer console and remove them:

For plugins:
```javascript
// Read all files in the virtual "plugins" directory:
require("fs").readdirSync("AppData/BetterDiscord/plugins");
// If you spot a file with a (1) or similar in the name, you can delete the file like this:
require("fs").unlinkSync("AppData/BetterDiscord/plugins/BetterRoleColors.plugin (1).js");
// Now reload the page and check!
```

For themes:
```javascript
// Read all files in the virtual "themes" directory:
require("fs").readdirSync("AppData/BetterDiscord/themes");
// If you spot a file with a (1) or similar in the name, you can delete the file like this:
require("fs").unlinkSync("AppData/BetterDiscord/themes/Catppuccin.theme (1).css");
// Now reload the page and check!
```

**Reply by tomboylover93** [reply]
Decided to test this to make sure that the file naming conventions were the cause of the problem and sure enough, they were.

I also saw you added a fix to this issue so I'll update the browser extension, that pretty much solves the problem I had. Thanks for your help.
